cshake.go

```diff
@@ -0,0 +1,215 @@
+//go:build !cmd_go_bootstrap && (cgo || goexperiment.ms_nocgo_opensslcrypto)
+
+package openssl
+
+import (
+	"runtime"
+	"strconv"
+
+	"github.com/golang-fips/openssl/v2/internal/ossl"
+)
+
+// shakeOneShot applies the SHAKE extendable output function to data and
+// writes the output to out.
+func shakeOneShot(secuirtyBits int, data []byte, out []byte) {
+	// Can't use EVP_Digest because it doesn't support output lengths
+	// larger than the block size, while crypto/sha3 supports any length.
+	alg := loadShake(secuirtyBits)
+	if alg == nil {
+		panic("openssl: unsupported SHAKE" + strconv.Itoa(secuirtyBits) + " function")
+	}
+	ctx, err := ossl.EVP_MD_CTX_new()
+	if err != nil {
+		panic(err)
+	}
+	defer ossl.EVP_MD_CTX_free(ctx)
+	if _, err := ossl.EVP_DigestInit_ex(ctx, alg.md, nil); err != nil {
+		panic(err)
+	}
+	if _, err := ossl.EVP_DigestUpdate(ctx, data); err != nil {
+		panic(err)
+	}
+	if _, err := ossl.EVP_DigestFinalXOF(ctx, out, len(out)); err != nil {
+		panic(err)
+	}
+}
+
+// SumSHAKE128 applies the SHAKE128 extendable output function to data and
+// returns an output of the given length in bytes.
+func SumSHAKE128(data []byte, length int) []byte {
+	out := make([]byte, length)
+	shakeOneShot(128, data, out[:])
+	return out
+}
+
+// SumSHAKE256 applies the SHAKE256 extendable output function to data and
+// returns an output of the given length in bytes.
+func SumSHAKE256(data []byte, length int) []byte {
+	out := make([]byte, length)
+	shakeOneShot(256, data, out[:])
+	return out
+}
+
+// SupportsSHAKE returns true if the SHAKE extendable output functions
+// with the given securityBits are supported.
+func SupportsSHAKE(securityBits int) bool {
+	if vMajor == 1 || (vMajor == 3 && vMinor < 3) {
+		// SHAKE MD's are supported since OpenSSL 1.1.1,
+		// but EVP_DigestSqueeze is only supported since 3.3,
+		// and we need it to implement [sha3.SHAKE].
+		return false
+	}
+	return loadShake(securityBits) != nil
+}
+
+// SupportsCSHAKE returns true if the CSHAKE extendable output functions
+// with the given securityBits are supported.
+func SupportsCSHAKE(securityBits int) bool {
+	return false
```

Comment:
Add a link to the OpenSSL tracker issue for CSHAKE: https://github.com/openssl/openssl/issues/28358